Source/x11/XGServerWindow.m

```diff
@@ -2707,62 +2692,109 @@ - (void) setWindowdevice: (int)win forContext: (NSGraphicsContext *)ctxt
 static Pixmap xIconMask;
 static BOOL didCreatePixmaps;
 
--(int) _createAppIconPixmaps
+Pixmap
+image_mask(Display *xdpy, Drawable draw, const unsigned char *data,
+           int w, int h, int colors, int alpha_treshold)
+{
+  int j, i;
+  unsigned char	ialpha;
+  Pixmap pix;
+  int bitmapSize = ((w + 7) >> 3) * h; // (w/8) rounded up times height
+  char *aData = calloc(1, bitmapSize);
+  char *cData = aData;
+
+  if (colors == 4)
+    {
+      int k;
+      for (j = 0; j < h; j++)
+	{
+	  k = 0;
+	  for (i = 0; i < w; i++, k++)
+	    {
+	      if (k > 7)
+		{
+	      	  cData++;
+	      	  k = 0;
+	      	}
+	      data += 3; // skip R, G, B
+	      ialpha = (unsigned short)((char)*data++);
+	      if (ialpha > alpha_treshold)
+                {
+		  *cData |= (0x01 << k);
+                }
+	    }
+	  cData++;
+	}
+    }
+  else
+    {
+      for (j = 0; j < bitmapSize; j++)
+	{
+	  *cData++ = 0xff;
+	}
+    }
+
+  pix = XCreatePixmapFromBitmapData(xdpy, draw, (char *)aData, w, h,
+				    1L, 0L, 1);
+  free(aData);
+  return pix;
+}
+
+- (int) _createAppIconPixmaps
 {
-  NSImage *image;
   NSBitmapImageRep *rep;
-  int i, j, w, h, samples, screen;
-  unsigned char *data;
-  XColor pixelColor;
-  GC pixgc;
-  RColor pixelRColor;
-  RContext *rcontext;
+  int              i, j, width, height, samples, screen;
+  unsigned char    *data;
+  RContext         *rcontext;
+  RXImage          *rxImage;
+  char             *r, *g, *b, *a;
 
   NSAssert(!didCreatePixmaps, @"called _createAppIconPixmap twice");
 
   didCreatePixmaps = YES;
 
-  image = [NSApp applicationIconImage];
-  rep = getStandardBitmap(image);
+  rep = getStandardBitmap([NSApp applicationIconImage]);
   if (rep == nil)
     return 0;
 
-  data = [rep bitmapData];
   screen = [[[self screenList] objectAtIndex: 0] intValue];
-  xIconPixmap = XCreatePixmap(dpy,
-                      [self xDisplayRootWindowForScreen: screen],
-                      [rep pixelsWide], [rep pixelsHigh],
-                      DefaultDepth(dpy, screen));
-  pixgc = XCreateGC(dpy, xIconPixmap, 0, NULL);
-
-  h = [rep pixelsHigh];
-  w = [rep pixelsWide];
-  samples = [rep samplesPerPixel];
   rcontext = [self xrContextForScreen: screen];
+  width = [rep pixelsWide];
+  height = [rep pixelsHigh];
+  samples = [rep samplesPerPixel];
 
-  for (i = 0; i < h; i++)
+  /**/
+  rxImage = RCreateXImage(rcontext, rcontext->depth, width, height);
+  xIconPixmap = XCreatePixmap(dpy, rcontext->drawable,
+                              width, height, rcontext->depth);
+  r = rxImage->image->data;
+  g = rxImage->image->data + 1;
+  b = rxImage->image->data + 2;
+  a = rxImage->image->data + 3;
+  data = [rep bitmapData];
+  for (i = 0; i < height; i++)
     {
       unsigned char *d = data;
-      for (j = 0; j < w; j++)
+      for (j = 0; j < width; j++)
 	{
-	  pixelRColor.red = d[0];
-	  pixelRColor.green = d[1];
-	  pixelRColor.blue = d[2];
-
-	  RGetClosestXColor(rcontext, &pixelRColor, &pixelColor);
```

Comment:
With this change the code won't run that well in indexed displays but I am not sure whether any GNUstep application was executed in such an environment in the last twenty years, so for me this change is fine.